### PR TITLE
Configure external Docusaurus theme

### PR DESCRIPTION
**Pull request by lunelson**
This PR sets up Docusaurus to depend on `@rasahq/docusaurus-theme-tabula`, under development in https://github.com/RasaHQ/tabula

Staging preview: https://staging--rasahq-docsv2-rasa.netlify.app/docs/rasa/next/

closes https://github.com/RasaHQ/growth/issues/1501

### CODE REVIEW

**Comment by lunelson** [comment]
@m-vdb could you take a look at this, as a sort of "pre-review"? I'd like to merge it tomorrow and then start a new branch so as not to have toooo big of a PR 😅; I have one outstanding problem though which is that ReDoc is failing webpack...

**Comment by m-vdb** [comment]
yes sure, should we convert the PR from a draft to a real PR?

File: docs/.prettierrc
Comment (m-vdb) [inline docs/.prettierrc:5]: ouhhhhh, I would greatly disagree here ^^

File: docs/TODO.md
Comment (m-vdb) [inline docs/TODO.md:11]: do we need to version this file?

File: docs/docs/action-server.mdx
Comment (m-vdb) [inline docs/docs/action-server.mdx:9]: is this intended? nothing in header

File: docs/docs/http-api-spec.mdx
Comment (m-vdb) [inline docs/docs/http-api-spec.mdx:8]: is this intended? nothing in header

File: docs/docusaurus.config.js
Comment (m-vdb) [inline docs/docusaurus.config.js:27]: I've merged a fix for this in master

File: docs/docusaurus.config.js
Comment (m-vdb) [inline docs/docusaurus.config.js:158]: to remove?

File: docs/docusaurus.config.js
Comment (m-vdb) [inline docs/docusaurus.config.js:162]: it's really hard to review because of the space to tabs change. AFAIK we use spaces everywhere in Rasa and Rasa X. Can you change to spaces? 🙏 🙏 🙏 🙏

**Review by m-vdb — CHANGES_REQUESTED** [review]
@lunelson I cannot review this properly because of the tabs :/

File: docs/TODO.md
Comment (lunelson) [reply docs/TODO.md:11]: nope

File: docs/docs/action-server.mdx
Comment (lunelson) [reply docs/docs/action-server.mdx:9]: yes the header contains a link to the API page, this is where redoc will live, so it can occupy the full page width

File: docs/docs/http-api-spec.mdx
Comment (lunelson) [reply docs/docs/http-api-spec.mdx:8]: as above

File: docs/docusaurus.config.js
Comment (lunelson) [reply docs/docusaurus.config.js:158]: I'll put it back in actually. I haven't quite figured out whether that option is good or not, I commented it out yesterday when I couldn't figure out the webpack errors

File: docs/docusaurus.config.js
Comment (lunelson) [reply docs/docusaurus.config.js:162]: Yes I will do it although I changed my position on this due to the accessiblity argument 🤷

Comment (lunelson) [reply docs/docusaurus.config.js:162]: I was intending to use tabs in the new design system...

File: docs/docs/action-server.mdx
Comment (lunelson) [reply docs/docs/action-server.mdx:9]: My intention is to flag these files for @erohmensing i.e. that they need to be updated or removed

File: docs/docusaurus.config.js
Comment (m-vdb) [reply docs/docusaurus.config.js:162]: please no 🙏 😄

File: docs/docusaurus.config.js
Comment (lunelson) [reply docs/docusaurus.config.js:27]: OK I'll revert this

File: docs/docs/action-server.mdx
Comment (m-vdb) [reply docs/docs/action-server.mdx:9]: but in the meantime, it means that the deployed doc won't have the API spec available. What about simply adding a comment in the header for Ella? Something like `FIXME: @erohmensing to handle this change` or something like that?

File: docs/.prettierrc
Comment (lunelson) [reply docs/.prettierrc:5]: reverted

File: docs/docusaurus.config.js
Comment (lunelson) [reply docs/docusaurus.config.js:27]: I've changed this to match master, where these two properties have the same value. I don't think that's the intention of the tagline 🤔  though

File: docs/docs/action-server.mdx
Comment (lunelson) [reply docs/docs/action-server.mdx]: OK reverted this too. Will ping ella when the new solution is up

**Comment by m-vdb** [comment]
@lunelson can you merge master in your branch? 🙏

**Comment by lunelson** [comment]
@erohmensing just FYI, this is a first pass on design adjustments that @eliseboyd and I have been discussing and wireframing, including changes to navigation. 

Part of this is considering the UX of moving the start page out of the nested structure and leaving the categories collapsed:

image

File: docs/docs/prototype-an-assistant.mdx
Comment (m-vdb) [inline docs/docs/prototype-an-assistant.mdx:7]: I don't think it's a good idea to rely on global imports instead of importing exactly what we need in each files. I would agree that maybe for `<Tabs>` there is a benefit (it's standard, it's from docusaurus, it's used in a lot of places), but for the case of the Prototyper, and surely other components, I think it's a bad idea and doesn't really scale.

File: docs/docusaurus.config.js
Comment (m-vdb) [inline docs/docusaurus.config.js:36]: I think that a lot of people liked the dark mode - any reason why we do not keep it around?

File: docs/docusaurus.config.js
Comment (m-vdb) [inline docs/docusaurus.config.js:76]: should we use `path.join()` here too?

File: docs/src/_theme/DocSidebar/index.jsx
Comment (m-vdb) [inline docs/src/_theme/DocSidebar/index.jsx:1]: any reason why the folder in named `_theme` instead of `theme`?

File: docs/src/_theme/DocSidebar/index.jsx
Comment (m-vdb) [inline docs/src/_theme/DocSidebar/index.jsx:17]: `Ph` is a bit cryptic to me. Do we need to keep the name because of sizzling? what does it stand for?

File: docs/src/_theme/DocSidebar/index.jsx
Comment (m-vdb) [inline docs/src/_theme/DocSidebar/index.jsx:191]: I'm guessing it would be tackled separately?

File: docs/src/_theme/DocSidebar/index.jsx
Comment (m-vdb) [inline docs/src/_theme/DocSidebar/index.jsx:250]: can we get the content for this before merging to master?

File: docs/src/_theme/DocSidebar/index.jsx
Comment (m-vdb) [inline docs/src/_theme/DocSidebar/index.jsx:255]: is there a specific TODO here?

Comment (m-vdb) [reply docs/src/_theme/DocSidebar/index.jsx:255]: (same for the one below)

File: docs/src/_theme/DocSidebar/index.jsx
Comment (m-vdb) [inline docs/src/_theme/DocSidebar/index.jsx:2]: is this file generated by sizzling? if so, should we mention it somewhere in the docstring of the file?

File: docs/src/_theme/DocSidebar/styles.module.scss
Comment (m-vdb) [inline docs/src/_theme/DocSidebar/styles.module.scss:6]: should we mentioned that this is generated by sizzling?

File: docs/src/components/prototyper/download-button.jsx
Comment (m-vdb) [inline docs/src/components/prototyper/download-button.jsx:1]: I am not sure I follow the logic of adding this to the theme, instead of having it in `src/components/`. To me, what is in the theme is usable by different pages. In case of the prototyper, it's not the case

File: docs/src/components/redoc.jsx
Comment (m-vdb) [inline docs/src/components/redoc.jsx:10]: there are tabs in this file

File: docs/src/stylesheets/resets/_destyle.scss
Comment (m-vdb) [inline docs/src/stylesheets/resets/_destyle.scss:1]: any reason why it's all on one line?

File: docs/src/stylesheets/theme/alerts.scss
Comment (m-vdb) [inline docs/src/stylesheets/theme/alerts.scss:1]: let's exclude stylecheets from typo CI (see `.typo-ci.yml` file at the root of the repo)

File: docs/src/utils/Ph/index.js
Comment (m-vdb) [inline docs/src/utils/Ph/index.js:6]: what does Ph stand for? can we have a minimal docstring?

File: docs/.eslintrc.js
Comment (m-vdb) [inline docs/.eslintrc.js:1]: are any of prettier / eslintrc checks run during CI?

**Review by m-vdb — CHANGES_REQUESTED** [review]
amazing effort @lunelson 💪 🎉 

I've added some comments to the code and I think some of them will need to be addressed before it's merge-able. Some other general comments that I think we should discuss:

- lots of scss files in `src/stylesheets` and fonts in `src/fonts`: is there any way to install those via yarn?
- the amount of theme files in huge and I am not sure it's a good idea to add them to our master right now: can we move this to a separate open source repository and install from git? (until we figure out a better way to make it available via Github packages or something else). CCing @tmbo on this, to double check if you agree with the approach.
- `make install-docs` failed with the following, so I couldn't test:

```bash
[4/4] 🔨  Building fresh packages...
error /Users/max/Sources/rasa/docs/node_modules/fibers: Command failed.
Exit code: 127
Command: node build.js || nodejs build.js
Arguments:
Directory: /Users/max/Sources/rasa/docs/node_modules/fibers
Output:
gyp info it worked if it ends with ok
gyp info using node-gyp@5.0.7
gyp info using node@13.11.0 | darwin | x64
gyp info find Python using Python version 2.7.16 found at "/System/Library/Frameworks/Python.framework/Versions/2.7/Resources/Python.app/Contents/MacOS/Python"
gyp info spawn /System/Library/Frameworks/Python.framework/Versions/2.7/Resources/Python.app/Contents/MacOS/Python
gyp info spawn args [
gyp info spawn args   '/usr/local/Cellar/node/13.11.0/libexec/lib/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'make',
gyp info spawn args   '-I',
gyp info spawn args   '/Users/max/Sources/rasa/docs/node_modules/fibers/build/config.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/usr/local/Cellar/node/13.11.0/libexec/lib/node_modules/npm/node_modules/node-gyp/addon.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/Users/max/Library/Caches/node-gyp/13.11.0/include/node/common.gypi',
gyp info spawn args   '-Dlibrary=shared_library',
gyp info spawn args   '-Dvisibility=default',
gyp info spawn args   '-Dnode_root_dir=/Users/max/Library/Caches/node-gyp/13.11.0',
gyp info spawn args   '-Dnode_gyp_dir=/usr/local/Cellar/node/13.11.0/libexec/lib/node_modules/npm/node_modules/node-gyp',
gyp info spawn args   '-Dnode_lib_file=/Users/max/Library/Caches/node-gyp/13.11.0/<(target_arch)/node.lib',
gyp info spawn args   '-Dmodule_root_dir=/Users/max/Sources/rasa/docs/node_modules/fibers',
gyp info spawn args   '-Dnode_engine=v8',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
gyp info spawn args   '--generator-output',
gyp info spawn args   'build',
gyp info spawn args   '-Goutput_dir=.'
gyp info spawn args ]
No receipt for 'com.apple.pkg.CLTools_Executables' found at '/'.

No receipt for 'com.apple.pkg.DeveloperToolsCLILeo' found at '/'.

No receipt for 'com.apple.pkg.DeveloperToolsCLI' found at '/'.

gyp: No Xcode or CLT version detected!
gyp ERR! configure error
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (/usr/local/Cellar/node/13.11.0/libexec/lib/node_modules/npm/node_modules/node-gyp/lib/configure.js:351:16)
gyp ERR! stack     at ChildProcess.emit (events.js:315:20)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:275:12)
gyp ERR! System Darwin 19.4.0
gyp ERR! command "/usr/local/Cellar/node/13.11.0/bin/node" "/usr/local/Cellar/node/13.11.0/libexec/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild" "--release"
gyp ERR! cwd /Users/max/Sources/rasa/docs/node_modules/fibers
gyp ERR! node -v v13.11.0
gyp ERR! node-gyp -v v5.0.7
gyp ERR! not ok
node-gyp exited with code: 1
Please make sure you are using a supported platform and node version. If you
would like to compile fibers on this machine please make sure you have setup your
build environment--
Windows + OS X instructions here: https://github.com/nodejs/node-gyp
Ubuntu users please run: `sudo apt-get install g++ build-essential`
RHEL users please run: `yum install gcc-c++` and `yum groupinstall 'Development Tools'`
Alpine users please run: `sudo apk add python make g++`
/bin/sh: nodejs: command not found
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
make: *** [install-docs] Error 127
```

**Comment by erohmensing** [comment]
Is there a reason why we should have them completely collapsed when there is so much space for discoverability? When talking about it with @karen-white, she thought it was very helpful to have the first level of `Building Assisants` `Concepts` and APIs open so that people have a general idea of where they'll find what type of information without having to open stuff up.

**Comment by lunelson** [comment]
@erohmensing yeah this is just a tryout, and I see that point for sure. I think what was irritating us there, was that there's little visual differentiation, neither between levels nor between categories vs pages, so there's more styling work to be done there

File: docs/docs/prototype-an-assistant.mdx
Comment (lunelson) [reply docs/docs/prototype-an-assistant.mdx:7]: I agree the Prototyper is probably too large to import on every page, yes, but importing the common UI elements like Tabs, Buttons etc. is a fairly standard MDX pattern. Will adjust

File: docs/docusaurus.config.js
Comment (lunelson) [reply docs/docusaurus.config.js:36]: It's not within our design system, we don't have a dark-mode palette. It's already been warned about, but we intend to look at bringing it back.

Comment (lunelson) [reply docs/docusaurus.config.js:36]: CC @eliseboyd ^^

File: docs/docusaurus.config.js
Comment (lunelson) [reply docs/docusaurus.config.js:76]: I'll kill this whole line eventually, it's not stable to rely on hashed assets generated by gatsby, I'll bring `rasa-logo.svg` in to `/static/`

File: docs/src/_theme/DocSidebar/index.jsx
Comment (lunelson) [reply docs/src/_theme/DocSidebar/index.jsx:1]: To prevent it from overriding. I changed customization methods part-way through since the `swizzle` approach was proving tedious (I was having to make many small changes in different component files), but I'm keeping these two components here because I'd already started altering them. Shortly I'll salvage the changes from here and move them in to the theme

File: docs/src/_theme/DocSidebar/index.jsx
Comment (lunelson) [reply docs/src/_theme/DocSidebar/index.jsx:17]: `Ph` stands for "placeholder". It's a component I use in development to block out space. It'll be phased out.

File: docs/src/components/redoc.jsx
Comment (lunelson) [reply docs/src/components/redoc.jsx:10]: you got me

File: docs/src/stylesheets/resets/_destyle.scss
Comment (lunelson) [reply docs/src/stylesheets/resets/_destyle.scss:1]: It's just a reset, copied from https://nicolas-cusan.github.io/destyle.css/, that I usually use, but I might phase it out

File: docs/src/stylesheets/theme/alerts.scss
Comment (lunelson) [reply docs/src/stylesheets/theme/alerts.scss:1]: Nice catch. I'll exclude `*.jsx` too then if that's OK. In any case if I extract all this, it will be within the dependency

File: docs/src/utils/Ph/index.js
Comment (lunelson) [reply docs/src/utils/Ph/index.js:6]: Answered above 😬 😃

File: docs/.eslintrc.js
Comment (lunelson) [reply docs/.eslintrc.js:1]: Not as far as I know, those configs weren't otherwise here. But this too will be moved in to the extracted theme.

**Comment by lunelson** [comment]
>lots of scss files in src/stylesheets and fonts in src/fonts: is there any way to install those via yarn?

The scss files will ideally be extracted along with the theme, which means I'll have to combine sass support with the theme definition 🤔 

The font files are commercially licensed fonts not installable by yarn but might also be moved in to the theme if Docusaurus supports that mechanic (themes supplying files to `/static/`), not sure

> the amount of theme files in huge and I am not sure it's a good idea to add them to our master right now: can we move this to a separate open source repository and install from git?

Yeah I was having a similar thought earlier, that developing this on master seems a little too intense. The origin of this PR thought was to try to break all this up a bit and review/merge in pieces. I wasn't confident before about Docusaurus relying on an external theme but it feels like it should work now. I'll give it a try, a github or github-pkg install should work

> make install-docs failed with the following, so I couldn't test

I think you're having the same `node-gyp` issues you had before, probably due to the `fibers` package.

File: docs/src/stylesheets/theme/alerts.scss
Comment (lunelson) [reply docs/src/stylesheets/theme/alerts.scss:1]: ...although, I notice that `*.md` files are specifically excluded too. I wonder what's it actually linting then? You probably want to *not* ignore `*.md` and `*.mdx` no?

File: docs/docusaurus.config.js
Comment (m-vdb) [reply docs/docusaurus.config.js:36]: ok 👍

**Review by m-vdb — CHANGES_REQUESTED** [review]
nice work @lunelson 🚀 

I still can't install the docs locally:

Screenshot 2020-08-25 at 10 00 11

**Comment by lunelson** [comment]
>I still can't install the docs locally

@m-vdb Yes, you'll need to authorize yourself with a global `~/.npmrc` file 😄. I've just documented this in [this Notion page](https://www.notion.so/rasa/GitHub-Package-Registry-9660161f04db45b392b3a6b160508b66)

**Comment by m-vdb** [comment]
Thanks for the doc, Lu! really helpful. One thing I am not sure about: do we really need this package to be private? Setting up a `~/.npmrc` file could be another friction in the installation process for contributors, so I'm wondering if we could simply have it public. (cc @tmbo WDYT?)

**Comment by m-vdb** [comment]
@lunelson when running the docs locally:
- `install-docs` runs properly (after I setup my npmrc)
- `livedocs` runs properly
- I am able to load the index page
- ⛔ I am not able to view any other page of the site
- ⛔ there are a lot of errors in the console

File: docs/docusaurus.config.js
Comment (m-vdb) [inline docs/docusaurus.config.js:66]: what should be the tagline?

File: docs/sidebars.js
Comment (m-vdb) [inline docs/sidebars.js:2]: at the bottom of the sidebar I see "TODO: ADD ITEMS HERE", it seems to be in the theme itself:
Screenshot 2020-09-08 at 14 27 12

File: docs/docusaurus.config.js
Comment (m-vdb) [inline docs/docusaurus.config.js:3]: i'd rename them `rehypeThemePlugins` and `remarkThemePlugins` respectively, as it's more explicit when reading the config file

File: docs/docusaurus.config.js
Comment (m-vdb) [inline docs/docusaurus.config.js:15]: is this still WIP?

File: docs/docusaurus.config.js
Comment (m-vdb) [inline docs/docusaurus.config.js:94]: should we have `SITE_URL` here?

File: docs/docusaurus.config.js
Comment (m-vdb) [inline docs/docusaurus.config.js:27]: nothing is done with this variable

File: docs/docs/installation.mdx
Comment (m-vdb) [inline docs/docs/installation.mdx:175]: Is the red effect intended when hovering the button?

Screenshot 2020-09-08 at 14 36 32

File: docs/docs/prototype-an-assistant.mdx
Comment (m-vdb) [inline docs/docs/prototype-an-assistant.mdx:7]: just to be sure I understand: why is it under the `themes/` folder instead of the `components/` folder?

**Review by m-vdb — CHANGES_REQUESTED** [review]
Thanks for the updates Lu 🚀 I've added a couple of comments to address before it is mergeable.

Another thing that we have to figure out: the theme is currently private and cannot be installed by anyone without some config. It's detrimental to the developer UX and also requires updates for our CI pipelines (if you look at the GH action runs, it's unable to install the dependency).

Can we:
- move the GH repository to public
- and add a visible notice on the README to say that it's alpha and intended for internal use at the moment
?

🙏

File: docs/docusaurus.config.js
Comment (lunelson) [reply docs/docusaurus.config.js]: I'll do it in the next one, if that's OK. It will be internalized in the theme.

File: docs/sidebars.js
Comment (lunelson) [reply docs/sidebars.js:2]: Yes, there are some points which come after the menu, including the link for the API docs. Current design looks like this. These things will be solved before launch or removed

image

File: docs/docusaurus.config.js
Comment (lunelson) [reply docs/docusaurus.config.js:3]: 👍 can do

File: docs/docusaurus.config.js
Comment (lunelson) [reply docs/docusaurus.config.js:15]: Yes, it will be solved before launch

File: docs/docusaurus.config.js
Comment (lunelson) [reply docs/docusaurus.config.js:94]: `SWAP_URL == SITE_URL` in production 🤷

File: docs/docusaurus.config.js
Comment (lunelson) [reply docs/docusaurus.config.js:27]: not yet

File: docs/docs/prototype-an-assistant.mdx
Comment (lunelson) [reply docs/docs/prototype-an-assistant.mdx:7]: I can't quite remember now. It seemed like the `@theme` alias was more stable than `@site`; but anyway I found it clearer to have customized and overridden components in one spot

File: docs/docs/installation.mdx
Comment (lunelson) [reply docs/docs/installation.mdx:175]: This is a placeholder for now, the styling isn't complete.

File: docs/sidebars.js
Comment (m-vdb) [reply docs/sidebars.js:2]: can we remove the TODO then? Otherwise it will land on the docs for the alpha, once merged into master. We cannot merge it like that :/

File: docs/docusaurus.config.js
Comment (m-vdb) [reply docs/docusaurus.config.js:97]: it's just that it doesn't work in dev.

File: docs/docs/prototype-an-assistant.mdx
Comment (m-vdb) [reply docs/docs/prototype-an-assistant.mdx:7]: it makes contributions and later updates to the theme more complicated for developers who do not know how docusaurus work, IMO. But not blocking, we can always move it back if we need to.

**Comment by m-vdb** [comment]
@lunelson it's still not merge-able. The items I think need to be addressed:
- make the theme public
- remove the user-visible placeholders (TODOs and such)
- the tagline in latin

I'm a bit concerned to merge this because in my opinion, we lose at least one feature: the possibility to support different versions. I know you plan on working on this, but I'd be cautious about promises. How long do you think handling the versions would take?

**Review by m-vdb — APPROVED** [review]
@lunelson 🚀 🚀 🚀 🚀 🚀  this is it! it's finally in a merge-able state. Thanks for your help with debugging yarn/npm stuff.

**Pull request merged**
